### PR TITLE
Merge Preparation: Merge Precise Object Tagging to EOP 32-Bit

### PR DESCRIPTION
**Pull request by luc-blaeser**
Prepare merging https://github.com/dfinity/motoko/pull/4544 to https://github.com/dfinity/motoko/pull/4193 (level 1 of PR stack)

### CODE REVIEW

File: src/codegen/compile.ml [hunk 1]
```diff
@@ -4091,7 +4120,7 @@ module Object = struct
         List.sort compare in
       let hash_blob =
         let hash_payload = StaticBytes.[ i32s hashes ] in
-        Blob.constant env (StaticBytes.as_bytes hash_payload)
+        Blob.constant env Tagged.T (StaticBytes.as_bytes hash_payload)
```
Comment (luc-blaeser) [inline src/codegen/compile.ml:4123]:
```suggestion
        Blob.constant env Tagged.B (StaticBytes.as_bytes hash_payload)
```

Comment (luc-blaeser) [reply src/codegen/compile.ml:4123]: This is the wrong tag.

File: src/codegen/compile.ml [hunk 2]
```diff
@@ -6267,19 +6312,19 @@ module Serialization = struct
...
       let descriptor = get_global_type_descriptor env in
-      Blob.load_data_segment env E.(descriptor.candid_data_segment) (get_candid_data_length env)
+      Blob.load_data_segment env Tagged.B E.(descriptor.candid_data_segment) (get_candid_data_length env)
     )
 
   let get_global_type_offsets env =
     Tagged.share env (fun env -> 
       let descriptor = get_global_type_descriptor env in
-      Blob.load_data_segment env E.(descriptor.type_offsets_segment) (get_type_offsets_length env)
+      Blob.load_data_segment env Tagged.T E.(descriptor.type_offsets_segment) (get_type_offsets_length env)
```
Comment (luc-blaeser) [inline src/codegen/compile.ml:6321]:
```suggestion
      Blob.load_data_segment env Tagged.B E.(descriptor.type_offsets_segment) (get_type_offsets_length env)
```

File: src/codegen/compile.ml [hunk 3]
```diff
@@ -6267,19 +6312,19 @@ module Serialization = struct
...
     )
 
   let get_global_idl_types env =
     Tagged.share env (fun env -> 
       let descriptor = get_global_type_descriptor env in
-      Blob.load_data_segment env E.(descriptor.idl_types_segment) (get_idl_types_length env)
+      Blob.load_data_segment env Tagged.T E.(descriptor.idl_types_segment) (get_idl_types_length env)
```
Comment (luc-blaeser) [inline src/codegen/compile.ml:6327]:
```suggestion
      Blob.load_data_segment env Tagged.B E.(descriptor.idl_types_segment) (get_idl_types_length env)
```

File: src/codegen/compile.ml [hunk 4]
```diff
@@ -8389,8 +8436,8 @@ module Persistence = struct
     let (candid_type_desc, type_offsets, type_indices) = Serialization.(type_desc env Persistence [actor_type]) in
     let serialized_offsets = StaticBytes.(as_bytes [i32s (List.map Int32.of_int type_offsets)]) in
     assert (type_indices = [0l]);
-    Blob.lit env candid_type_desc ^^
-    Blob.lit env serialized_offsets ^^
+    Blob.lit env Tagged.B candid_type_desc ^^
+    Blob.lit env Tagged.T serialized_offsets ^^
```
Comment (luc-blaeser) [inline src/codegen/compile.ml:8440]:
```suggestion
    Blob.lit env Tagged.B serialized_offsets ^^
```

File: rts/motoko-rts/src/persistence.rs [hunk 5]
```diff
@@ -192,7 +192,8 @@ pub unsafe fn register_stable_type<M: Memory>(
     new_candid_data: Value,
     new_type_offsets: Value,
 ) {
-    assert_eq!(new_candid_data.tag(), TAG_BLOB);
+    assert_eq!(new_candid_data.tag(), TAG_BLOB_B);
+    assert_eq!(new_type_offsets.tag(), TAG_BLOB_T);
```
Comment (luc-blaeser) [inline rts/motoko-rts/src/persistence.rs:196]:
```suggestion
    assert_eq!(new_type_offsets.tag(), TAG_BLOB_B);
```

**Comment by luc-blaeser** [comment]
> LGTM. Checked some of the additional commits on top too.

Thank you very much!

**Comment by crusso** [comment]
(Still LGTM)

**Pull request merged**